metadata-ingestion/src/datahub/ingestion/source/dremio/dremio_source.py

```diff
@@ -0,0 +1,616 @@
+                            f"Adding source_type {source_type} to mapping as database. Error: {exc}"
+                        )
+
+                        self.reference_source_mapping.add_mapping(
+                            source_type, source_name
+                        )
+                        dremio_source_type = self.reference_source_mapping.get_category(
+                            source_type
+                        )
+
+                    source_map[source_platform_name.lower()] = DremioSourceMapping(
+                        platform=source_type,
+                        platform_name=source_name,
+                        dremio_source_type=dremio_source_type,
+                    )
+
+            else:
+                logger.error(
+                    f'Source "{source.container_name}" is broken. Containers will not be created for source.'
+                )
+                logger.error(
+                    f'No new cross-platform lineage will be emitted for source "{source.container_name}".'
+                )
+                logger.error("Fix this source in Dremio to fix this issue.")
+
+        return source_map
+
+    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Generate workunits for Dremio metadata.
+        """
+
+        self.source_map = self._build_source_map()
+
+        for wu in self.get_workunits_internal():
+            self.report.report_workunit(wu)
+            yield wu
+
+        # Emit the stale entity removal workunits
+        yield from self.stale_entity_removal_handler.gen_removed_entity_workunits()
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Internal method to generate workunits for Dremio metadata.
+        """
+
+        # Process Containers
+        containers = self.dremio_catalog.get_containers()
+        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
+            future_to_container = {
+                executor.submit(self.process_container, container): container
+                for container in containers
+            }
+
+            for future in as_completed(future_to_container):
+                container_info = future_to_container[future]
+                try:
+                    yield from future.result()
+                    logger.info(
+                        f"Dremio container {container_info.container_name} emitted successfully"
+                    )
+                except Exception as exc:
+                    self.report.num_containers_failed = +1
+                    self.report.report_failure(
+                        "Failed to process Dremio container",
+                        f"Failed to process container {'.'.join(container_info.path)}.{container_info.resource_name}: {exc}",
+                    )
+
+        # Process Datasets
+        datasets = self.dremio_catalog.get_datasets()
+
+        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:

```

Comment:
Also, why are we using multithreading here ?

Reply:
Not needed removed